riscv: implement maskisr steponly command

erhankur opened the pull request:
With this PR,  `riscv maskisr <off | steponly>` command implemented. Now it will be possible to disable/restore interrupts within the step function.  

ARM implementation of this command has also `auto and on` option but it is not ready for the riscv yet.  

I will also run and share the test results to make sure at least nothing broken. 

Change-Id: I1a3b666d466b064460c3acc307a36485ce165601
Signed-off-by: Erhan Kurubas <erhan.kurubas@espressif.com>

erhankur committed:
riscv: implement maskisr steponly command

Change-Id: I1a3b666d466b064460c3acc307a36485ce165601
Signed-off-by: Erhan Kurubas <erhan.kurubas@espressif.com>

timsifive commented on src/target/riscv/riscv.c:
```diff
@@ -2916,6 +2915,31 @@ COMMAND_HANDLER(riscv_use_bscan_tunnel)
+	}
+
+	n = jim_nvp_value2name_simple(nvp_maskisr_modes, info->isrmask_mode);
+	command_print(CMD, "riscv interrupt mask %s", n->name);
```
Is that expected? I don't think any of the riscv commands echo their result. Should they?

erhankur replied:
Printing result allows us to get the current value without setting.  It can be useful for the tests like `store_existing_setting, change_it, restore_previous_value` 

`riscv maskisr` --> will return existing value

Also I think it is good to see the result in the telnet command line.

JanMatCodasip replied:
Another option for your consideration: You may display the current setting only if the command is called without any parameters. For example OpenOCD's commands `tcl_port`, `telnet_port` (and some others) work this way. Unfortunately, OpenOCD commands do not behave consistently and we would find exceptions. 

In any case, it would be good to remain consistent at least within the riscv target. I would like to learn what is the preferred behavior from the folks in upstream, then update the riscv commands to follow that.

erhankur replied:
Ok. changed the command as `set_maskisr` and did not print information for set.

timsifive commented on src/target/riscv/riscv.h:
```diff
@@ -56,6 +56,13 @@ enum riscv_halt_reason {
 };
 
+enum riscv_isrmasking_mode {
+	RISCV_ISRMASK_AUTO,	/* not supported yet */
```
Let's not include them in the enum at all then. If you want them listed as a future reminder, then comment them out.

erhankur replied:
I left comment for the future.

timsifive reviewed (changes requested):
Overall the change looks OK to me. Can you also add documentation for the command in doc/openocd.texi?

JanMatCodasip commented on src/target/riscv/riscv.c:
```diff
@@ -3500,6 +3533,55 @@ static int riscv_resume_go_all_harts(struct target *target)
+{
+	RISCV_INFO(info);
+	uint64_t current_mstatus;
+	uint8_t mstatus_bytes[8] = { 0 };
```
Just a little thing: You can move the variable definitions closer to the point of their first use. 

OpenOCD's code style in upstream no longer requires (nor encourages) the old ANSI C style where variables had to be defined at the beginning of a block.

JanMatCodasip commented on src/target/riscv/riscv.c:
```diff
@@ -2372,12 +2356,26 @@ int riscv_openocd_step(struct target *target, int current,
 	}
 
+	if (info->isrmask_mode == RISCV_ISRMASK_STEPONLY) {
+		if (riscv_interrupts_restore(target, current_mstatus) != ERROR_OK)
+			return ERROR_FAIL;
+	}
+
```
Even if riscv_step_rtos_hart() fails, I believe we shall still try to restore mstatus to the original value, to leave the hart in as unchanged as possible state.

erhankur replied:
@JanMatCodasip thanks to point out this behavior. Looking at the other parts,  I was in between restore/return or just return. So what about enabling triggers if there is step or irq_restore  error?

erhankur replied:
@timsifive your comments are also welcome.

JanMatCodasip replied:
> Looking at the other parts, I was in between restore/return or just return.

You're right - the risk of not restoring everything has already been present in the original code (restore of triggers would not be attempted if step fails).

> So what about enabling triggers if there is step or irq_restore error?

IMO, an ideal solution would be to rework this function so that in case of any failure in any of the steps, whatever got changed gets restored.  I understand that this is additional work, not directly related to the new "maskisr" command.

Please create an issue for that so that we don't forget to return to this.

erhankur replied:
Didn't want to create another technical debt.  I did some restore work. Please check it again.

JanMatCodasip commented on src/target/riscv/riscv.c:
```diff
@@ -1952,26 +1952,11 @@ static int riscv_run_algorithm(struct target *target, int num_mem_params,
+	if (riscv_interrupts_disable(target, ie_mask, &current_mstatus) != ERROR_OK)
+		return ERROR_FAIL;
 
 	/* Run algorithm */
```
There is potential issue in riscv_run_algorithm() - if the algorithm execution fails at some point below this line (`return <whatever>;`), the saved register contents won't be restored - not even attempted.

This risk was already present in the original code, so I believe the pull request need not be blocked by that. The issue can be addressed in a separate patch.

erhankur replied:
Yes right. I just changed the existing code block with the function. Do you have a backlog to follow todo issues?

JanMatCodasip replied:
I have recorded the issue here: https://github.com/riscv/riscv-openocd/issues/682

JanMatCodasip commented on src/target/riscv/riscv.c:
```diff
@@ -3325,6 +3349,13 @@ static const struct command_registration riscv_exec_command_handlers[] = {
 			"1: DATA_REGISTER}"
 	},
+	{
+		.name = "maskisr",
```
The other option-setting TCL commands so far follow the `set_<something>` naming convention.

@timsifive - I do not have any preference about the naming convention of the commands but I'd like them to be unified. Similarly, it would be good to have a unified way how to just display the current value without setting a new one (see my earlier comment in this pull request). What is your opinion on that?

timsifive replied:
I agree on both points. Consistency is more important than the specific solution. Unfortunately we're pretty far away from that.

It looks like most riscv commands that set an option use set_, so let's stick with that for this PR.

As far as displaying a value, I prefer to show the value when there are no arguments, and to be silent when an option is set.

erhankur replied:
Done

JanMatCodasip commented on src/target/riscv/riscv.c:
```diff
@@ -3325,6 +3349,13 @@ static const struct command_registration riscv_exec_command_handlers[] = {
 	},
+	{
+		.name = "maskisr",
+		.handler = riscv_mask_interrupts,
```
Please consider renaming the handler function to `riscv_maskisr` to follow the naming convention already used for the other handlers.

JanMatCodasip commented on src/target/riscv/riscv.c:
```diff
@@ -1952,26 +1952,11 @@ static int riscv_run_algorithm(struct target *target, int num_mem_params,
-
-	reg_mstatus->type->get(reg_mstatus);
-	current_mstatus = buf_get_u64(reg_mstatus->value, 0, reg_mstatus->size);
 	uint64_t ie_mask = MSTATUS_MIE | MSTATUS_HIE | MSTATUS_SIE | MSTATUS_UIE;
```
Just a suggestion: Could you rename `ie_mask` to `irq_disabled_mask`, `disabled_irqs` (or something similar) - to make it clear that the bits indicate which irqs get disabled, not enabled. 

There are more occurences of `ie_mask`. Preferably, address them all.

JanMatCodasip reviewed (commented):
@erhankur - Thank you for your contribution.

Overall, the change looks all right. Please look at my feedback:

- two items are important: restoring of registers when a certain operation fails (single-step or algorithm)
- the remaining comments relate to cosmetic stuff like naming conventions

erhankur committed:
riscv: restore triggers and irq mask inside step function

Change-Id: I4e1b0665f4f2f75e42a6191c61634bdfa19ae2fb
Signed-off-by: Erhan Kurubas <erhan.kurubas@espressif.com>

JanMatCodasip commented on src/target/riscv/riscv.c:
```diff
@@ -2372,22 +2356,38 @@ int riscv_openocd_step(struct target *target, int current,
-	target->state = TARGET_HALTED;
-	target->debug_reason = DBG_REASON_SINGLESTEP;
-	target_call_event_callbacks(target, TARGET_EVENT_HALTED);
+_exit_riscv_openocd_step:
```
Cosmetic stuff: The goto labels have function scope, so you can safely use a shorter name - e.g. `done` or `exit`. There won't be any conflicts with labels in different functions.

erhankur replied:
Yes. Right. Used _exit.

JanMatCodasip commented on src/target/riscv/riscv.c:
```diff
@@ -2372,22 +2356,38 @@ int riscv_openocd_step(struct target *target, int current,
-	target->debug_reason = DBG_REASON_SINGLESTEP;
-	target_call_event_callbacks(target, TARGET_EVENT_HALTED);
+_exit:
+	out |= enable_triggers(target, trigger_state);
```
Please, do not use `|=` operator on the error codes. Although it works in this case (because ERROR_OK equals to zero), it is not clean. It may also lead to returning a non-sense error code, created as a bitwise mixture of two different error codes (`ERROR_SOMETHING1 | ERROR_SOMETHING2`).

One possible way how to fix it: use `bool success = true;` instead of the `int out` variable, set it to false when any operation fails. Then at the end of function: `return success ? ERROR_OK : ERROR_FAIL`.

If you decide to remember the full error code of the first failed call (instead of just success flag), that is also perfectly fine, but requires few more lines of code.

erhankur replied:
@JanMatCodasip thanks for feedbacks and reviews. Changed as you suggested. Please review again.

JanMatCodasip reviewed (commented):
Thank you for the changes. I inspected the code visually and it looks all right.

I have one last comment related to the code style (the use of `|=`).

timsifive commented:
Broader question about this PR. The debug spec defines dcsr.stepie, which defaults to 0. When it is 0 "Interrupts (including NMI) are disabled during single stepping." Why is this change needed? Is there hardware that has the bit wired to 1?

JanMatCodasip reviewed (commented):
The code looks all right, thanks.

Please, discuss also with Tim whether there is a specific reason why `dcsr.stepie` bit cannot be used. It did not occur to me initially that there is already this dedicated bit in `dcsr` for disabling the interrupts during step.

erhankur commented:
@JanMatCodasip thanks. I will clarify `stepie` bit usage. 

Can you also help me to pass the tests? No idea why `MemorySampleSingle` test fails. 

./gdbserver.py ./targets/RISC-V/spike32-2.py MemorySampleSingle

[20220218-084040-spike32_2-MemorySampleMixed.log](https://github.com/riscv/riscv-openocd/files/8095980/20220218-084040-spike32_2-MemorySampleMixed.log)

Warn : Error writing to GDB socket. Dropping the connection. 

It is not related to my branches. Same with `riscv`

erhankur commented:
OK. I will remove `riscv set_maskisr` command and dependencies. Can we keep the other parts? I can make another PR for the remaining part.

erhankur commented:
@timsifive As I just learned we implemented v0.13 for all riscv chips. Unfortunately `stepie` bit is not exist at v0.13. 

So, can we add this command to support earlier version of v0.13.2?

timsifive commented:
I thought that on specs before stepie was added, interrupts are disabled during single stepping. 0.13 is a work-in-progress version, and there are probably multiple variants of it floating around. Can you share the PDF or a URL of the exact document you implemented?

erhankur commented:
I have asked for the exact document. In the meantime, I found this https://static.dev.sifive.com/riscv-debug-spec-0.13.b4f1f43.pdf

timsifive commented:
Thanks. In the link you posted, the definition of the `step` bit says:

> When set and not in Debug Mode, the hart will
> only execute a single instruction and then enter
> Debug Mode. *Interrupts are disabled when this
> bit is set.* If the instruction does not complete due
> to an exception, the hart will immediately enter
> Debug Mode before executing the trap handler,
> with appropriate exception registers set.

erhankur commented:
Confirmed with the team. We implemented the url I sent above.

As a conclusion;
1 - We didn't implement `stepie` since it is not exist in v0.13
2 - Interrupts are kept `enabled` during a single step to give control to the debugger software. 

In my opinion, safest place to disable/enable interrupts is inside `riscv_openocd_step`.  Changes will stay at local,  no events will be fired without restoring, tested for the long time, working well etc.

I moved disable/enable routines to the upper layer  and run some of our tests but not worked as expected. Need to look at deeply. 

@timsifive sorry for bringing this on the table again.  I understand your concern but don't think so this will bring much overhead since `isrmask_mode` will be disabled by default.  Ordinary users will not touch this setting.  

So, can we add this command to the upstream? Otherwise looks like things will be getting more complex.

timsifive commented:
So to recap, you have hardware that does not implement stepie (because it's based on an old spec that doesn't have it) and keeps interrupts enabled during single step, even though the spec says they should be disabled.

stepie was defined so that older implementations would appear to have stepie hard-wired to 0, and everything just works. Except your hardware looks like stepie is hard-wired to 0 but interrupts are enabled in any case.

Right now OpenOCD doesn't allow the user to set stepie to 1. So long-term the functionality we want is:
1. Give the user a setting that says "step into interrupts" or "don't step into interrupts." OpenOCD can set stepie based on this setting. OpenOCD can detect whether the setting is supported, and give an error if the desired behavior isn't supported.
2. Give the user a setting that says "mask interrupts while stepping" which is orthogonal to the first one. This is really only useful if you do not want to step into interrupts, and the hardware doesn't support stepie=0.

This change implements the second part of that, so that's a good improvement.

Can you update the documentation to make it clear that this feature is only useful on hardware that always steps into interrupts, and also that disabling the new option does not guarantee that single stepping will go into interrupt handlers? (To make that happen, stepie would have to be written to 1 as well.)

erhankur committed:
doc: update for riscv set_maskisr command

Change-Id: Ia7d3a6df846cfc4568d79558f719e93f038aee9b
Signed-off-by: Erhan Kurubas <erhan.kurubas@espressif.com>

erhankur commented:
@timsifive thank you for your support. Document is updated.

JanMatCodasip commented on doc/openocd.texi:
```diff
@@ -10424,6 +10424,13 @@ tunneled DR scan consists of:
 
+@deffn {Command} {riscv set_maskisr} [@option{off}|@option{steponly}]
+Selects whether interrupts will be disabled when single stepping. The default configuration is @option{off}.
+This feature is only useful on hardware that always steps into interrupts and doesn't support stepie=0.
+Keep in mind, disabling the option does not guarantee that single stepping will go into interrupt handlers.
+To make that happen, stepie would have to be written to 1 as well.
```
Just one last cosmetic suggestion: For clarity, it would be good to mention also the register name - `stepie` --> `dcsr.stepie`.

When dealing with RISC-V debug, it is important to be very clear about which registers we talk about, because there are (at least?) three different register spaces involved:

* JTAG registers (IR, DRs)
* Debug module registers (dmstatus, dmcontrol, ...)
* Hart's achitectural registers (GPRs, CSRs, ...)

Thanks.

```suggestion
This feature is only useful on hardware that always steps into interrupts and doesn't support dcsr.stepie=0.
Keep in mind, disabling the option does not guarantee that single stepping will go into interrupt handlers.
To make that happen, dcsr.stepie would have to be written to 1 as well.
```

erhankur replied:
Done

JanMatCodasip replied:
Thanks.
